Not compatible with ESLint v9

ESLint 9 has [just been released](https://eslint.org/blog/2024/04/eslint-v9.0.0-released/), but this package is not yet compatible:

```
npm ERR! Could not resolve dependency:
npm ERR! peer eslint@"^6 || ^7 || ^8" from @codeyourfuture/eslint-config-standard@4.0.3
npm ERR! node_modules/@codeyourfuture/eslint-config-standard
npm ERR!   dev @codeyourfuture/eslint-config-standard@"^4.0.3" from the root project
```

## Workaround

Explicitly install `eslint@8` until this package is updated to support v9.

--- a/.eslintrc.root
+++ b/.eslintrc.root
@@ -1,7 +1,0 @@
-{
-  "env": {
-    "es6": true,
-    "node": true
-  },
-  "extends": "."
-}

--- a/.github/workflows/push.yml
+++ b/.github/workflows/push.yml
@@ -30,7 +30,7 @@ jobs:
     needs: build
     strategy:
       matrix:
-        eslint: [6, 7, 8, 9]
+        eslint: [8.27, 8, 9]
         node: [16, 18, 20]
         exclude:
           # ESLint v9 requires Node ^18.18.0 || ^20.9.0 || >=21.1.0
@@ -56,7 +56,7 @@ jobs:
     needs: build
     strategy:
       matrix:
-        eslint: [8, 9]
+        eslint: [8.27, 8, 9]
         node: [16, 18, 20]
         exclude:
           # ESLint v9 requires Node ^18.18.0 || ^20.9.0 || >=21.1.0

--- a/.github/workflows/push.yml
+++ b/.github/workflows/push.yml
@@ -51,10 +51,36 @@ jobs:
         ESLINT_USE_FLAT_CONFIG: 'false'
         ESLINT_VERSION: ${{ matrix.eslint }}
         PACK: 'false'
+  test-flat-config:
+    runs-on: ubuntu-22.04
+    needs: build
+    strategy:
+      matrix:
+        eslint: [8, 9]
+        node: [16, 18, 20]
+        exclude:
+          # ESLint v9 requires Node ^18.18.0 || ^20.9.0 || >=21.1.0
+          - eslint: 9
+            node: 16
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-node@v4
+        with:
+          cache: 'npm'
+          node-version: ${{ matrix.node }}
+      - uses: actions/download-artifact@v4
+        with:
+          name: package
+      - run: npm ci
+      - run: npm run test:install
+        env:
+          ESLINT_USE_FLAT_CONFIG: 'true'
+          ESLINT_VERSION: ${{ matrix.eslint }}
+          PACK: 'false'
   deploy-github:
     runs-on: ubuntu-22.04
     if: startsWith(github.ref, 'refs/tags/v')
-    needs: test-eslintrc
+    needs: [test-eslintrc, test-flat-config]
     steps:
       - uses: actions/checkout@v4
       - uses: actions/download-artifact@v4
@@ -70,7 +96,7 @@ jobs:
   deploy-npm:
     runs-on: ubuntu-22.04
     if: startsWith(github.ref, 'refs/tags/v')
-    needs: test-eslintrc
+    needs: [test-eslintrc, test-flat-config]
     steps:
     - uses: actions/checkout@v4
     - uses: actions/setup-node@v4

--- a/.github/workflows/push.yml
+++ b/.github/workflows/push.yml
@@ -25,13 +25,17 @@ jobs:
         if-no-files-found: error
         name: package
         path: codeyourfuture-eslint-config-standard-*.tgz
-  install:
+  test-eslintrc:
     runs-on: ubuntu-22.04
     needs: build
     strategy:
       matrix:
-        eslint: [6, 7, 8]
+        eslint: [6, 7, 8, 9]
         node: [16, 18, 20]
+        exclude:
+          # ESLint v9 requires Node ^18.18.0 || ^20.9.0 || >=21.1.0
+          - eslint: 9
+            node: 16
     steps:
     - uses: actions/checkout@v4
     - uses: actions/setup-node@v4
@@ -44,12 +48,13 @@ jobs:
     - run: npm ci
     - run: npm run test:install
       env:
+        ESLINT_USE_FLAT_CONFIG: 'false'
         ESLINT_VERSION: ${{ matrix.eslint }}
         PACK: 'false'
   deploy-github:
     runs-on: ubuntu-22.04
     if: startsWith(github.ref, 'refs/tags/v')
-    needs: install
+    needs: test-eslintrc
     steps:
       - uses: actions/checkout@v4
       - uses: actions/download-artifact@v4
@@ -65,7 +70,7 @@ jobs:
   deploy-npm:
     runs-on: ubuntu-22.04
     if: startsWith(github.ref, 'refs/tags/v')
-    needs: install
+    needs: test-eslintrc
     steps:
     - uses: actions/checkout@v4
     - uses: actions/setup-node@v4

--- a/README.md
+++ b/README.md
@@ -30,15 +30,33 @@ Install this package along with ESLint itself:
 npm install --save-dev eslint @codeyourfuture/eslint-config-standard
 ```
 
-Then create an [ESLint config file] and add this config to the `"extends"` section:
+Then create an [ESLint config file] and add this config:
+
+```javascript
+const cyfConfig = require("@codeyourfuture/eslint-config-standard");
+
+module.exports = [cyfConfig];
+```
+
+or using ES module syntax:
+
+```javascript
+import cyfConfig from "@codeyourfuture/eslint-config-standard";
+
+export default [cyfConfig];
+```
+
+Alternatively, for a slightly more permissive set of rules, you can use `@codeyourfuture/eslint-config-standard/lax`.
+
+### `.eslintrc`
+
+If you have not yet migrated to the newer ESLint "flat config", you can apply these rules to the [deprecated config] using `"extends"`:
 
 ```json
 {
   "extends": ["@codeyourfuture/standard"]
 }
 ```
-
-Alternatively, for a slightly more permissive set of rules, you can extend `@codeyourfuture/eslint-config-standard/lax`.
 
 ## Principles
 
@@ -89,7 +107,8 @@ You can clone this repo and run `npm install` to install the development depende
   [brace-style]: https://eslint.org/docs/rules/brace-style
   [comma-dangle]: https://eslint.org/docs/rules/comma-dangle
   [curly]: https://eslint.org/docs/rules/curly
-  [ESLint config file]: https://eslint.org/docs/user-guide/configuring
+  [deprecated config]: https://eslint.org/docs/latest/use/configure/configuration-files-deprecated
+  [ESLint config file]: https://eslint.org/docs/latest/use/configure/configuration-files
   [indent]: https://eslint.org/docs/rules/indent
   [linebreak-style]: https://eslint.org/docs/rules/linebreak-style
   [no-trailing-spaces]: https://eslint.org/docs/rules/no-trailing-spaces

--- a/README.md
+++ b/README.md
@@ -60,9 +60,9 @@ If you have not yet migrated to the newer ESLint "flat config", you can apply th
 
 ## Principles
 
- 1. **Errors only** - don't train students to ignore *any* output, all rules should either be `"error"` or `"off"`
+ 1. **Errors only** - don't teach trainees to ignore *any* output, all rules should either be `"error"` or `"off"`
  2. **Maximise consistency** - where there are options (e.g. braces for single-line statements, parentheses around arrow function parameters), be consistent with the non-optional cases
- 3. **Minimise change set size** - keep commits small so students can focus on the important changes
+ 3. **Minimise change set size** - keep commits small so trainees can focus on the important changes
 
 ## Rules
 
@@ -82,7 +82,7 @@ This config starts from [`eslint:recommended`][1] then adds the following rules:
 | standard, lax | [object-curly-spacing] | `"always"` | |
 | standard, lax | [operator-linebreak] | `"before"` | |
 | standard, lax | [quotes] | `"double", { "avoidEscape": true, "allowTemplateLiterals": false }` | More likely to need `'` inside a string than `"` |
-| standard, lax | [semi] | | Students shouldn't have to memorise the [ASI rules] |
+| standard, lax | [semi] | | Trainees shouldn't have to memorise the [ASI rules] |
 
 ## Development
 

--- a/README.md
+++ b/README.md
@@ -66,7 +66,7 @@ If you have not yet migrated to the newer ESLint "flat config", you can apply th
 
 ## Rules
 
-This config starts from [`eslint:recommended`][1] then adds the following rules:
+This config starts from [`js.configs.recommended`][1] then adds the following rules:
 
 | Configuration| Rule | Setting | Principles/rationale |
 |---|---|---|---|
@@ -99,7 +99,7 @@ You can clone this repo and run `npm install` to install the development depende
     no version conflicts and lints `index.js`. E.g. `ESLINT_VERSION=6 npm run test` will test that this configuration
     works with the latest version of ESLint 6.
 
-  [1]: https://eslint.org/docs/user-guide/configuring#using-eslintrecommended
+  [1]: https://eslint.org/docs/latest/use/configure/configuration-files#using-predefined-configurations
   [2]: https://www.reddit.com/r/javascript/comments/c8drjo/nobody_talks_about_the_real_reason_to_use_tabs/
 
   [arrow-parens]: https://eslint.org/docs/rules/arrow-parens

--- a/bin/examples.sh
+++ b/bin/examples.sh
@@ -2,20 +2,23 @@
 set -uo pipefail
 
 HERE="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
-EXAMPLES="$HERE/../examples"
-ESLINT="$HERE/../node_modules/.bin/eslint"
+EXAMPLES="$(cd "$HERE/../examples" && pwd)"
+ESLINT="$(cd "$HERE/../node_modules/.bin" && pwd)/eslint"
 
 for DIR in $EXAMPLES/*; do
-  echo "checking $DIR"
+  FLAT="$([[ -f "$DIR/eslint.config.js" ]] && echo 'true' || echo 'false')"
+  CONFIG="$DIR/$([[ "$FLAT" = 'true' ]] && echo 'eslint.config.js' || echo '.eslintrc')"
 
-  if "$ESLINT" "$DIR/pass.js"; then
+  echo "checking $DIR (flat=$FLAT)"
+
+  if ESLINT_USE_FLAT_CONFIG="$FLAT" "$ESLINT" --config "$CONFIG" "$DIR/pass.js"; then
     echo "$DIR pass.js success"
   else
     echo "$DIR pass.js failure"
     exit 1
   fi
 
-  if "$ESLINT" "$DIR/fail.js"; then
+  if ESLINT_USE_FLAT_CONFIG="$FLAT" "$ESLINT" --config "$CONFIG" "$DIR/fail.js"; then
     echo "$DIR fail.js failure"
     exit 1
   else

--- a/bin/test.sh
+++ b/bin/test.sh
@@ -6,7 +6,7 @@ set -euo pipefail
 echo "Testing with ESLint v$ESLINT_VERSION"
 HERE="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 PACK="${PACK:-}"
-ROOT_DIR="$HERE/.."
+ROOT_DIR="$(cd "$HERE/.." && pwd)"
 TEST_DIR="$ROOT_DIR/test"
 rm -rf "$TEST_DIR"
 
@@ -41,6 +41,34 @@ cat > "$TEST_DIR/package.json" <<- EndOfMessage
   }
 }
 EndOfMessage
+
+cat > "$TEST_DIR/.eslintrc" <<- EndOfMessage
+{
+  "env": {
+    "es6": true,
+    "node": true
+  },
+  "extends": "@codeyourfuture/eslint-config-standard/lax",
+  "parserOptions": {
+    "ecmaVersion": 9
+  }
+}
+EndOfMessage
+
+cat > "$TEST_DIR/eslint.config.js" <<- EndOfMessage
+const { node } = require("globals");
+const cyfConfig = require("@codeyourfuture/eslint-config-standard/lax");
+module.exports = [
+  {
+    languageOptions: {
+      ecmaVersion: 9,
+      globals: node,
+    },
+  },
+  cyfConfig,
+];
+EndOfMessage
+
 cp $ROOT_DIR/*.js $TEST_DIR
 
 # Make sure ESLint and the config install and lint without issues
@@ -50,7 +78,8 @@ pushd "$TEST_DIR"
   npm ls --depth 0  # see https://stackoverflow.com/a/63177495/3001761
 
   npm run lint -- --version
-  npm run lint -- *.js
+  ESLINT_USE_FLAT_CONFIG=false npm run lint -- *.js
+  ESLINT_USE_FLAT_CONFIG=true npm run lint -- *.js
 popd
 
 # Tidy up

--- a/eslint.root.js
+++ b/eslint.root.js
@@ -1,0 +1,19 @@
+const { node } = require("globals");
+
+const cyfConfig = require(".");
+
+/** @type {import("eslint").Linter.FlatConfig} */
+module.exports = [
+	{
+		languageOptions: {
+			ecmaVersion: 9,
+			globals: node,
+		},
+	},
+	cyfConfig,
+	{
+		rules: {
+			"quote-props": ["error", "consistent-as-needed"],
+		},
+	},
+];

--- a/examples/.eslintrc
+++ b/examples/.eslintrc
@@ -1,6 +1,0 @@
-{
-  "env": {
-    "es6": true,
-    "node": true
-  }
-}

--- a/examples/flat/eslint.config.js
+++ b/examples/flat/eslint.config.js
@@ -1,0 +1,14 @@
+const { node } = require("globals");
+
+const cyfConfig = require("../..");
+
+/** @type {import("eslint").Linter.FlatConfig} */
+module.exports = [
+  {
+    languageOptions: {
+      ecmaVersion: 9,
+      globals: node,
+    },
+  },
+  cyfConfig,
+];

--- a/examples/flat/fail.js
+++ b/examples/flat/fail.js
@@ -1,0 +1,3 @@
+var thing = {
+  foo: 123
+}

--- a/examples/flat/pass.js
+++ b/examples/flat/pass.js
@@ -1,0 +1,5 @@
+const thing = {
+	foo: 123,
+};
+
+console.log(thing);

--- a/examples/lax/.eslintrc
+++ b/examples/lax/.eslintrc
@@ -3,5 +3,6 @@
     "es6": true,
     "node": true
   },
-  "extends": "../../lax"
+  "extends": "../../lax",
+  "root": true
 }

--- a/examples/lax/.eslintrc
+++ b/examples/lax/.eslintrc
@@ -1,3 +1,7 @@
 {
+  "env": {
+    "es6": true,
+    "node": true
+  },
   "extends": "../../lax"
 }

--- a/examples/standard/.eslintrc
+++ b/examples/standard/.eslintrc
@@ -1,3 +1,7 @@
 {
+  "env": {
+    "es6": true,
+    "node": true
+  },
   "extends": "../../."
 }

--- a/examples/standard/.eslintrc
+++ b/examples/standard/.eslintrc
@@ -3,5 +3,6 @@
     "es6": true,
     "node": true
   },
-  "extends": "../../."
+  "extends": "../../.",
+  "root": true
 }

--- a/index.js
+++ b/index.js
@@ -1,9 +1,10 @@
 "use strict";
+const lax = require("./lax");
 
 module.exports = {
-	extends: ["./lax"],
 	rules: {
-		"indent": ["error", "tab", { "SwitchCase": 1 }],
+		...lax.rules,
+		"indent": ["error", "tab", { SwitchCase: 1 }],
 		"linebreak-style": ["error", "unix"],
 	},
 };

--- a/lax.js
+++ b/lax.js
@@ -1,18 +1,19 @@
 "use strict";
+const { configs: { recommended } } = require("@eslint/js");
 
 module.exports = {
-	extends: ["eslint:recommended"],
 	rules: {
+		...recommended.rules,
 		"arrow-parens": "error",
-		"brace-style": ["error", "1tbs", { "allowSingleLine": false }],
+		"brace-style": ["error", "1tbs", { allowSingleLine: false }],
 		"comma-dangle": ["error", "always-multiline"],
 		"curly": "error",
 		"no-trailing-spaces": "error",
-		"no-unused-vars": ["error", { "ignoreRestSiblings": true }],
+		"no-unused-vars": ["error", { ignoreRestSiblings: true }],
 		"no-var": "error",
 		"object-curly-spacing": ["error", "always"],
 		"operator-linebreak": ["error", "before"],
-		"quotes": ["error", "double", { "avoidEscape": true, "allowTemplateLiterals": false }],
+		"quotes": ["error", "double", { avoidEscape: true, allowTemplateLiterals: false }],
 		"semi": "error",
 	},
 };

--- a/package.json
+++ b/package.json
@@ -11,7 +11,7 @@
   ],
   "main": "index.js",
   "scripts": {
-    "lint": "eslint --config .eslintrc.root *.js",
+    "lint": "eslint --config eslint.root.js *.js",
     "test:examples": "./bin/examples.sh",
     "test:install": "./bin/test.sh"
   },
@@ -36,6 +36,6 @@
     "eslint": "^8.44.0"
   },
   "peerDependencies": {
-    "eslint": "^6 || ^7 || ^8 || ^9"
+    "eslint": "^8.27 || ^9"
   }
 }

--- a/package.json
+++ b/package.json
@@ -36,6 +36,6 @@
     "eslint": "^8.44.0"
   },
   "peerDependencies": {
-    "eslint": "^6 || ^7 || ^8"
+    "eslint": "^6 || ^7 || ^8 || ^9"
   }
 }

--- a/package.json
+++ b/package.json
@@ -33,7 +33,7 @@
   "homepage": "https://github.com/codeyourfuture/eslint-config-standard#readme",
   "funding": "https://github.com/sponsors/CodeYourFuture",
   "devDependencies": {
-    "eslint": "^8.44.0"
+    "eslint": "^9.1.1"
   },
   "peerDependencies": {
     "eslint": "^8.27 || ^9"

